src/google/protobuf/compiler/csharp/csharp_field_base.h

```diff
@@ -61,7 +61,9 @@ class FieldGeneratorBase : public SourceGeneratorBase {
   virtual void GenerateMembers(io::Printer* printer) = 0;
   virtual void GenerateMergingCode(io::Printer* printer) = 0;
   virtual void GenerateParsingCode(io::Printer* printer) = 0;
+  virtual void GenerateBufferParsingCode(io::Printer* printer);
```

Comment:
all methods are abstract (`=0`), should the new ones be too?

Reply:
@JunTaoLuo

Reply:
There is a default implementation for these virtual methods https://github.com/protocolbuffers/protobuf/pull/5888/files#diff-468783f0b0e231e04220b6f6efe59aa1R164-R171

Reply:
Reopening this item - still needs addressing IMHO.

Reply:
The only difference between the parsing code is that, we'll need to pass the `CodedInputReader` and `CodedOutputWriter` by reference, for example: https://github.com/protocolbuffers/protobuf/pull/5888/files#diff-9ec70ca68be79446ca22c0eb97bec0a2R90. For many of the types, there's no difference between the parsing code: https://github.com/protocolbuffers/protobuf/blob/2941b08aee57004e74b933df29803ed2e3237875/src/google/protobuf/compiler/csharp/csharp_primitive_field.cc#L309.

We can certainly make `GenerateBufferParsingCode` completely distinct from `GenerateParsingCode`. The downside is that we'll have more code duplication and we may forget to copy over changes from one codepath to the other. The upside is that we it's less likely that we accidentally generate the wrong buffer parsing code due to the silent fallback.

Reply:
Any preferences @jtattermusch ?

Reply:
@JunTaoLuo  thanks for the explanation. I found that the only files where the parsing/serialization code stays the same is csharp_message_field.cc and csharp_primitive_field.cc. Because in majority of cases we need to provide implementation, I changed the methods in the base class to abstract (=0)  and provided functionality explicitly in those two files (I pushed a commit with the fix to this PR).